Refactor [#121] writing diary 부분 리펙토링

kim-seonwoo opened the pull request:
## 🍀 작업 내용
<!-- 작업한 내용을 간단하게 적어주세요! -->
### 시행착오를 통해 생각한 부분
- RxCollectionViewSectionedReloadDataSources는 오로지 text만을 바라보고 있는 것이 좋다.
텍스트필드의 상태를 enum으로 정의하여 통합하려했지만, 이는 불필요하게 순환적인 렌더링을 발생시킨다고 생각하였습니다.
` viewModel.diaryTextsRelay` 만을 구독하여 이에 따른 변화만을 바라보게 하였습니다.
바라보고 있는 text에 대한 텍스트필드의 상태는 cell에서 text를 받아서 처리합니다.

### 텍스트필드 주요 흐름

VC의 configureCollectionView() 에서 텍스트필드의 흐름을 처리하고 있습니다.

``` swift
                cell.textView.rx.text.orEmpty
                    .skip(1)
                    .distinctUntilChanged()
                    .do(onNext: { [weak cell] text in
                        guard let cell = cell else { return }
                        
                        let limitedText = String(text.prefix(50))
                        if cell.textView.text != limitedText {
                            cell.textView.text = limitedText
                        }

                        cell.textInputLabel.text = "\(limitedText.count)"

                        let isValid = limitedText.count < 50
                        cell.limitErrorLabel.isHidden = isValid
                        cell.writingContainer.makeBorder(
                            width: 1,
                            color: isValid ? .mainYellow : .redCustom
                        )
                    })
                    .subscribe(onNext: { [weak self, weak cell] _ in
                        guard let self = self, let cell = cell else { return }
                        self.updateTextViewHeightIfNeeded(for: cell, collectionView)
                    })
                    .disposed(by: cell.disposeBag)
   ```
rx.text.orEmpty는 텍스트뷰를 계속 바라보면서 타이핑 하나하나 감지합니다.
50자 이상으로 작성방지,
텍스트 글자세기 업데이트,
텍스트 글자수 경고 UI 업데이트,
동적 높이 조정을 여기서 실행 합니다.

```swift
                cell.textView.rx.didBeginEditing
                    .subscribe(onNext: { [weak self, weak cell] in
                        guard let self = self, let cell = cell else { return }
                        
                        var flags = self.viewModel.isPlaceholderRelay.value
                        flags[indexPath.item] = false
                        self.viewModel.isPlaceholderRelay.accept(flags)
                        
                        cell.updateUIOnBeginEditing()
                    })
                    .disposed(by: cell.disposeBag)

````
텍스트 변경이 시작될때, 
isPlaceholderRelay를 업데이트 시켜 처음 수정한것이 아님을 업데이트해 줍니다.

```swift                
                cell.textView.rx.didEndEditing
                    .subscribe(onNext: { [weak self, weak cell] in
                        guard let self = self, let cell = cell else { return }
                        var items = self.viewModel.diaryTextsRelay.value
                        items[indexPath.item] = cell.textView.text
                        self.viewModel.diaryTextsRelay.accept(items)
                    })
                    .disposed(by: cell.disposeBag)
```
didEndEditing 에서만 오로지 diaryTextsRelay를 업데이트 해 줍니다.

```swift

                let isPlaceholder = self.viewModel.isPlaceholderRelay.value[indexPath.item]
                
                cell.bindData(index: indexPath.item + 1, text: text, isPlaceholder: isPlaceholder)
```

cell에 다음과 같은 정보를 넘겨줍니다.
textRelay가 업데이트될때 바인딩이 일어납니다.

## 🚀 PR Point
<!-- 주의할 사항이나 같이 고민해볼 부분, 강조하고 싶은 내용 등을 적어주세요! -->
- 이 방식이 최선의 방식인지...


## ✚ 코드 리뷰 반영 사항
<!-- 반영한 코드 리뷰 내용이 있다면 적어주세요! -->

## ✅ CheckList
- [ ] 오류 없이 빌드되는지 확인
- [ ] 로그용 print문 제거
- [ ] 불필요한 주석 제거
- [ ] 코드 컨벤션 확인

### 🔗 Issue 
<!-- 생성한 관련 이슈가 있다면 Resolved #이슈번호로 닫아주세요! -->
Resolved #121

kim-seonwoo committed:
[Fix/#121] 답장 불가 시간 홈으로 분기

kim-seonwoo committed:
[Feat/#115] 저장 시 텍스트뷰 포커스 끄기

kim-seonwoo committed:
Merge remote-tracking branch 'origin/feat/#118-draft-bottomsheet' into Refactor/#121-WritingDiary

kim-seonwoo committed:
[Fix/#121] empty String 변경

kim-seonwoo committed:
[Fix/#121] 불필요한 relay 삭제

kim-seonwoo committed:
[Delete/#121] limitFive Alert 삭제

kim-seonwoo committed:
[Refactor/#121] text.orEmpty 반복 합치기

kim-seonwoo committed:
[Delete/#121] 불필요한 주석 삭제

Nya128 commented on Clody_iOS/Clody_iOS/Presentation/WritingDiary/ViewModels/WritingDiaryViewModel.swift:
```diff
@@ -50,12 +48,8 @@ final class WritingDiaryViewModel: ViewModelType {
-    let isFirstRelay = BehaviorRelay<[Bool]>(value: [true])
-    let textDidEditing = PublishRelay<String>()
-    let textEndEditing = PublishRelay<String>()
+    let diaryTextsRelay = BehaviorRelay<[String]>(value: [""])
+    let isPlaceholderRelay = BehaviorRelay<[Bool]>(value: Array(repeating: true, count: 5))
```
계속 드는 생각은... 이 두개의 relay가 굳이 나뉘어져 있어야 하나? 싶네요. 🤔
어차피 둘다 일기 cell 개수만큼 배열이 있을 테니까, 두개보다 하나로 합쳐서 관리하는 게 유지보수도 편할듯 합니다.
일기 내용 String값과 placeholder 여부 Bool값 합쳐서 새로운 모델을 만드는 게 좋을 거 같슴다

kim-seonwoo replied:
placeholder, String, textview 상태를 합쳐서 만들어봤는데..
제가 했던 바로는 합친 relay로 하면 이 relay를 컬렉션뷰가 바라보고 있어서 과도하게 계속 바인딩되고,
루프적인 것이 발생하여 일단 단순화하여 분리하였습니다..

이 문제만 해결방법을 찾으면 모델로 만드는게 베스트일 거 같긴합니더, 텍스트뷰의 에러 상태까지도..

Nya128 commented on Clody_iOS/Clody_iOS/Presentation/WritingDiary/Cells/WritingDiaryCell.swift:
```diff
@@ -165,23 +165,60 @@ final class WritingDiaryCell: UICollectionViewCell {
-            writingContainer.backgroundColor = .grey09
-            writingContainer.makeBorder(width: 0, color: .clear)
-        } else {
+        let trimmedText = text.trimmingCharacters(in: .whitespacesAndNewlines)
```
오 양쪽 끝 공백 제거.. 좋습니다

Nya128 commented on Clody_iOS/Clody_iOS/Presentation/WritingDiary/ViewModels/WritingDiaryViewModel.swift:
```diff
@@ -127,23 +113,18 @@ final class WritingDiaryViewModel: ViewModelType {
-            .asDriver(onErrorJustReturn: [])
-        
-        let isFirst = isFirstRelay
+        let items: Driver<[WritingDiarySection]> = diaryTextsRelay
+            .map { models in
+                [WritingDiarySection(header: "Diary Header", items: models)]
+            }
```
궁금한 것이 있습니다!!
일기 작성 화면에서 사실상 최대 5개의 리스트가 전부라서 단일 섹션으로도 충분히 구현이 가능한 것 같아서요. 현재는 output의 items Driver에 'WritingDiarySection 배열' 타입으로 정의되어 있는데, 그럼 한개의 cell당 하나의 섹션인 거니까,, 불필요하게 구성된 느낌이 드는데 제가 생각한게 맞나요?
이게 약간 RxCollectionViewSectionedReloadDataSource를 억지로 짜맞추려 쓴 거 같이 보여서 지금 상황에서는 굳이 RxDataSource 안 써도 될거 같다는 생각..입니다. 오히려 쉽게 가도 될듯?

kim-seonwoo replied:
옙옙 맞습니다..
현재 그래서 단일 섹션으로 구현되고 있습니다.
헷갈릴 수 있는 부분이 [WritingDiarySection(header: "Diary Header", items: models)] 요부분의 타입으로 넘겨주는게 WritingDiarySection객체 하나당 하나의 섹션이고,
models가 실질적으로 cell에 바인딩된다고 할 수 있습니다.

간소화 한다면 지금 기능처럼 diaryTextsRelay를 바라보되 RxCollectionViewSectionedReloadDataSource를 안쓰는 방향으로는 할 수 있을것같습니다

Nya128 commented on Clody_iOS/Clody_iOS/Presentation/WritingDiary/ViewModels/WritingDiaryViewModel.swift:
```diff
@@ -127,23 +113,18 @@ final class WritingDiaryViewModel: ViewModelType {
+
         let showDraftAlert = input.tapBackButton.asSignal()
         
-        let isAddButtonEnabled = Observable.combineLatest(input.tapAddButton.asObservable(), diariesRelay.asObservable())
-            .map { _, diaries in diaries.count < 5 }
```
오... 세세하게 잘 수정했다

Nya128 committed:
[Refactor/#121] DiaryItem, DiaryState 모델 생성 및 UI 디테일 캡슐화

Nya128 committed:
[Refactor/#121] 일기작성 cell bindData() 함수에서 분기 처리 로직 DiaryItem 모델로 이동

Nya128 committed:
[Refactor/#121] RxDataSource -> 단일 섹션 rx items로 변경 및 diaryText, isPlaceholder relay -> diaryStateRelay 하나로 리팩토링

- RxCollectionViewSectionedReloadDataSource 삭제
- input viewDidLoad 삭제
- diaryStateRelay: 컬렉션뷰 데이터소스
- diaryTextBufferRelay: 최신 일기 데이터 상태 (계속 셀 바인딩되어 불필요한 컬렉션뷰 UI 업데이트 방지. 보내기/임시저장 시 diaryStateRelay로 현재값 방출. == 중간 역할)

Nya128 committed:
[Chore/#121] 코드 정리

- 불필요한 self 삭제
- 네이밍, 타입 변경
- 코드 순서 변경

Nya128 committed:
[Feat/#121] 답장 상세 화면 뒤로가기 탭할 시 홈화면으로 이동

kim-seonwoo committed:
[Fix/#121] 첫 cell 에러 수정

Nya128 committed:
[Fix/#121] 답장 상세 화면에서 바로 popToRootVC 할 경우, 답장 대기 화면 deinit 안 되는 이슈 해결

Nya128 committed:
[Chore/#121] 일기작성 cell 텍스트뷰 속성 추가 및 코드 정리

Nya128 committed:
[Fix/#121] 보내기 버튼 눌렀을 때 endEditing 시점, updateDiaryState() 호출 시점 수정

Nya128 committed:
[Fix/#121] 똑같은 DiaryState 객체를 참조함으로써 생긴 diaryStateRelay, bufferRelay 동기화 문제 해결 (DiaryState 모델 struct로 변경)

- diaryStateRelay.accept(currentBufferState)
: 위 코드는 DiaryState 모델이 class라서 레퍼런스 타입일 경우, diaryTextBufferRelay에서 방출한 DiaryState 인스턴스를 그대로 diaryStateRelay에 방출하는 것 -> 서로 똑같은 객체를 참조하게 됨.
분명 bufferRelay만 accept() 했어도 diaryStateRelay에도 동기화된 이유는 같은 객체를 수정한 것이기 때문.
- DiaryState 모델을 struct로 수정 -> 참조 아닌 copy 되도록 value type으로 변경

Nya128 committed:
[Fix/#121] 일기작성 텍스트가 changed 될 때마다 bufferRelay에 accept 되도록 수정

- 기존 didEndEditing에서만 updateTextBuffer() 호출하던 코드를 cell.textView.rx.text.orEmpty로 이동

Nya128 committed:
[Chore/#121] 일기작성 컬렉션뷰 바텀 contentInset 수정

=> (플러스버튼 높이 + 버튼과 키보드 사이 간격*2)

Nya128 committed:
[Fix/#121] trimmedText로 인해 입력한 텍스트가 사라지는 문제 해결

- text.trimmingCharacters(in: .whitespacesAndNewlines) 코드는 완전한 빈값인지, isEmpty 확인할 때만 사용

kim-seonwoo committed:
[Fix/#121] limitAlert 롤백

Nya128 reviewed (approved):
일기작성 리팩토링 어떻게든 하긴했다 ㅋㅋ 😂
저거 일기개수 5개 제한 Alert만 적절하게 잘 뜨기만 하면 될듯!

슬슬 스유로 넘어가봅시다 !!! 고생했으

kim-seonwoo committed:
[Feat/#121] limit toast error relay 추가

kim-seonwoo committed:
[Chore/#121] endEditing weak self 제거

kim-seonwoo committed:
[Fix/#121] 일기저장 Alert string 수정

kim-seonwoo committed:
[Fix/#121] Draft 상태 기존과 다를때만 DraftAlert

Nya128 commented on Clody_iOS/Clody_iOS/Presentation/WritingDiary/ViewModels/WritingDiaryViewModel.swift:
```diff
@@ -172,6 +173,10 @@ final class WritingDiaryViewModel: ViewModelType {
         diaryTextBufferRelay.accept(newState)
     }
+    
+    func isSameFromInitialDraft() -> Bool {
+        return currentDiaryState.items == initialDraftState.items
+    }
```
이거 함수말고 computed property로 정의하는거 어떤가요?
그리고 텍스트만 비교하도록 DiaryState의 getTexts() 함수 써봤어요
```
var isSameFromInitialDraft: Bool {
    currentDiaryState.getTexts() == initialDraftState.getTexts()
}
```

Nya128 commented on Clody_iOS/Clody_iOS/Presentation/WritingDiary/ViewControllers/WritingDiaryViewController.swift:
```diff
@@ -260,7 +260,7 @@ private extension WritingDiaryViewController {
                 viewModel.updateDiaryState()
                 
-                if !viewModel.hasAnyContent() {
+                if viewModel.isSameFromInitialDraft() {
```
오 네이밍 괜찮다. 🙂

Nya128 commented on Clody_iOS/Clody_iOS/Presentation/WritingDiary/Models/DiaryItem.swift:
```diff
@@ -7,7 +7,7 @@
+struct DiaryItem: Equatable {
     var text: String
     var isPlaceholder: Bool
     var isEditing: Bool = false
```
struct에 Equatable 다시 공부하고 왔어요 ☺️
이 상태에서는 그러면 text, isPlaceholder, isEditing 저 세개의 변수 모두 비교해서 같아야 true가 나올텐데 맞나요?
사실 임시저장에서는 text 배열만 가져가기 때문에 나머지 빼고 text 배열만 비교해도 무방하다고 생각합니다!
(만약 그렇게 한다면 DiaryItem에 Equatable 빼도 됨)

예를 들어 `isPlaceholder == true`인 텍스트뷰에 탭 한 번 했다고 false가 되면,, 다른 값이 되는거니까(어차피 둘다 텍스트 빈값인 건 똑같은데)

Nya128 reviewed (commented):
저거 구조체 비교 말고 일기 text 배열끼리만 비교하는 거 어떠신지... 😇
코멘트 봐주십숑 
빠른 작업.. 감사합니다 ㅎㅎ

kim-seonwoo committed:
[Fix/#121] 코드리뷰 반영: DraftAlert 관련

kim-seonwoo committed:
[Fix/#121] 버전 넘버 1.2.0으로 수정
